### PR TITLE
NamedTemporaryFile works different on Windows

### LINKED ISSUE #25: Permission denied for Temp file on Windows
When I run "pyannotate --type-info ./annotate.json ." with on Windows (both 2.7.14 and 3.6.3, probably others) , I get the following error:
Traceback (most recent call last):
  File "c:\python27\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\python27\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Python27\Scripts\pyannotate.exe\__main__.py", line 9, in <module>
  File "c:\python27\lib\site-packages\pyannotate_tools\annotations\__main__.py", line 45, in main
    generate_annotations_json(infile, tf.name)
  File "c:\python27\lib\site-packages\pyannotate_tools\annotations\main.py", line 58, in generate_annotations_json
    with open(target_path, 'w') as f:
IOError: [Errno 13] Permission denied: 'c:\\temp\\tmp2ui1ku'

A little bit of googling suggests this might be the problem [Permission Denied To Write To My Temporary File](https://stackoverflow.com/questions/23212435/permission-denied-to-write-to-my-temporary-file)

### PR DESCRIPTION
**Pull request by NeonGraal**
Add optional stream parameters to generate_annotation_json and parse_json.

Hopefully fixes #25 but unable to run all tests due to lib2to3 quirks on Windows and Ubuntu

### CODE REVIEW

File: pyannotate_tools/annotations/__main__.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def main():
 
     # Run pass 2 with output written to a temporary file.
     infile = args.type_info
-    generate_annotations_json(infile, tf.name)
+    generate_annotations_json(infile, tf.name, target_stream=tf.file)
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/__main__.py]: You don't need to use the .file attribute -- just pass tf, it's a file alright.

File: pyannotate_tools/annotations/main.py [hunk 2]
```diff
@@ -1,8 +1,9 @@
 """Main entry point to mypy annotation inference utility."""
 
 import json
+from io import IOBase
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/main.py]: I wouldn't do that. I'd import IO from typing and use IO[str] as the type of the streams everywhere.

File: pyannotate_tools/annotations/main.py [hunk 3]
```diff
@@ -22,16 +23,16 @@
                                           'samples': int})
 
 
-def generate_annotations_json(source_path, target_path):
-    # type: (str, str) -> None
+def generate_annotations_json(source_path, target_path, source_stream=None, target_stream=None):
+    # type: (str, str, Optional[IOBase], Optional[IOBase]) -> None
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/main.py]: IOBase -> IO[str]

File: pyannotate_tools/annotations/parse.py [hunk 4]
```diff
@@ -85,14 +86,17 @@ def __init__(self, comment):
...
 
     The input JSON is expected to to have a list of RawEntry items.
     """
-    with open(path) as f:
-        data = json.load(f)  # type: List[RawEntry]
+    if isinstance(stream, IOBase):
+        data = json.load(stream)
+    else:
+        with open(path) as f:
+            data = json.load(f)  # type: List[RawEntry]
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/parse.py]: The annotation should go on the first assignment to data.

File: pyannotate_tools/annotations/tests/main_test.py [hunk 5]
```diff
@@ -25,10 +25,12 @@ def test_generation(self):
             }
         ]
         """
-        target = tempfile.NamedTemporaryFile(mode='r')
-        with self.temporary_json_file(data) as source_path:
-            generate_annotations_json(source_path, target.name)
+        target = tempfile.NamedTemporaryFile(mode='w+')
+        with self.temporary_json_file(data) as source:
+            generate_annotations_json(source[0], target.name, source_stream=source[1], target_stream=target.file)
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/tests/main_test.py]: No need for .file here or anywhere.

File: pyannotate_tools/annotations/tests/main_test.py [hunk 6]
```diff
@@ -76,7 +78,8 @@ def test_ambiguous_kind(self):
     @contextlib.contextmanager
     def temporary_json_file(self, data):
         # type: (str) -> Iterator[str]
```
Comment (gvanrossum) [inline pyannotate_tools/annotations/tests/main_test.py]: Update the type here too (Iterator[Tuple[str, IO[str]]]).

File: .gitignore [hunk 7]
```diff
@@ -2,3 +2,5 @@
 /*.egg-info
 /build
 /dist
+.idea/
+*.pyc
```
[the remark this answers is not in the document]
Comment (NeonGraal) [reply .gitignore]: I'll take them out.

I've gotten in the habit at work of not presuming that people's global .gitignores are correct.

**Comment by NeonGraal** [comment]
I've signed the CLA now too

**Comment by gvanrossum** [comment]
Thanks! And fingers crossed for Windows. (I haven't managed to set up AppVeyor yet. Have you managed to test on Windows manually yet?)

**Comment by rowillia** [comment]
This PR appears to have broken Python 3 - 

```bash
$ pyannotate -w --type-info ./annotations.json $file
Traceback (most recent call last):
  File "/Users/rwilliams/src/oss/pyannotate/venv/bin/pyannotate", line 11, in <module>
    load_entry_point('pyannotate==1.0.2', 'console_scripts', 'pyannotate')()
  File "/Users/rwilliams/src/oss/pyannotate/venv/lib/python3.6/site-packages/pyannotate-1.0.2-py3.6.egg/pyannotate_tools/annotations/__main__.py", line 45, in main
  File "/Users/rwilliams/src/oss/pyannotate/venv/lib/python3.6/site-packages/pyannotate-1.0.2-py3.6.egg/pyannotate_tools/annotations/main.py", line 59, in generate_annotations_json
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/__init__.py", line 180, in dump
    fp.write(chunk)
  File "/Users/rwilliams/src/oss/pyannotate/venv/lib/python3.6/tempfile.py", line 483, in func_wrapper
    return func(*args, **kwargs)
TypeError: a bytes-like object is required, not 'str'
```

We should probably write a test that exercises `main`, perhaps just run through the example?

**Comment by gvanrossum** [comment]
Ouch. Yeah, we need a test for that! I'm submitting a revert as a PR (#42), but if you make a PR that fixes it (with a test) we can apply that instead.

**Comment by gvanrossum** [comment]
I'm going to revert it and come up with a way that doesn't require
temporary files at all.

**Pull request merged**